clientUtils/formatValue.ts

```diff
@@ -1,92 +1,199 @@
-                noSpaceUnit: shortNumberPrefixes,
-                numDecimalPlaces: 2,
-            })
-    } else if (!isNoSpaceUnit && shortNumberPrefixes && absValue >= 1e3) {
-        output = formatValue(value / 1e3, {
-            ...options,
-            unit: "k",
-            noSpaceUnit: true,
-            numDecimalPlaces: 2,
+function checkIsUnitCurrency(unit: string): unit is "$" | "£" {
+    return ["$", "£"].includes(unit)
+}
+
+function checkIsUnitPercent(unit: string): unit is "%" {
+    return unit[0] === "%"
+}
+
+function getTrim({ trailingZeroes }: { trailingZeroes: boolean }): "~" | "" {
+    return trailingZeroes ? "" : "~"
+}
+
+function getSign({ showPlus }: { showPlus: boolean }): "+" | "" {
+    return showPlus ? "+" : ""
+}
+
+function getSymbol({ unit }: { unit: string }): "$" | "" {
+    return checkIsUnitCurrency(unit) ? "$" : ""
+}
+
+function getType({
+    numberAbreviation,
+    value,
+}: {
+    numberAbreviation: "long" | "short" | false
+    value: number
+}): string {
+    // f: fixed-point notation (i.e. fixed number of decimal points)
+    // s: decimal notation with an SI prefix, rounded to significant digits
+    if (numberAbreviation === "long") {
+        // do not abbreviate until 1 million
+        return Math.abs(value) < 1e6 ? "f" : "s"
+    }
+    if (numberAbreviation === "short") {
+        // do not abbreviate until 1 thousand
+        return Math.abs(value) < 1000 ? "f" : "s"
+    }
+
+    return "f"
+}
+
+function getPrecision({
+    value,
+    numDecimalPlaces,
+    numberAbreviation,
+}: {
+    value: number
+    numDecimalPlaces: number
+    numberAbreviation: "short" | "long" | false
+}): string {
+    if (Math.abs(value) < 1e6 && numberAbreviation !== "short") {
+        return `${numDecimalPlaces}`
+    }
+
+    // when dealing with abreviated numbers, adjust precision so we get 12.84 million instead of 13 million
+    // the modulo one-liner counts the "place columns" of the number, resetting every 3
+    // 1 -> 1, 48 -> 2, 981 -> 3, 7222 -> 1
+    const numberOfDigits = String(Math.floor(Math.abs(value))).length
+    const precisionPadding = ((numberOfDigits - 1) % 3) + 1
+
+    // hard-coded 2 decimal places for abbreviated numbers
+    return `${precisionPadding + 2}`
+}
+
+function replaceSIPrefixes({
+    string,
+    numberAbreviation,
+}: {
+    string: string
+    numberAbreviation: "short" | "long"
+}): string {
+    const prefix = string[string.length - 1]
+
+    const prefixMap: Record<string, Record<string, string>> = {
+        short: {
+            k: "k",
+            M: "M",
+            G: "B",
+            T: "T",
+            P: "P",
+            E: "E",
+            Z: "Z",
+            Y: "Y",
+        },
+        long: {
+            k: "k",
+            M: " million",
+            G: " billion",
+            T: " trillion",
+            P: " quadrillion",
+            E: " quintillion",
+            Z: " sextillion",
+            Y: " septillion",
+        },
```

Comment:
Let's remove all after `trillion` (`T`)? I don't think they're widely understood.

Reply:
It's a bit tricky unfortunately. I agree, no one really knows these prefixes (hence remapping G to B) and there was work done in a [d3-format PR](https://github.com/d3/d3-format/pull/96) to introduce this functionality natively, but I don't think there's any out-of-the-box way to set a threshold on ab**b**reviation.

Some options we have:
- Switch to exponential notation after trillion (also handling cases where 9.5 trillion would get rounded to 1 quadrillion if precision is set to 1)
- Re-implement the logic to get it to output 1,234T (possibly quite hard to do with this flow)
- Change the prefixes from `P` to `kT` or `quad` or `thousand trillion` or whatever.

Thoughts?

Reply:
On the options:

- I don't think we should be switching to exponential notation in this PR. As we discussed privately, when using exponential notation all values should be in exponential notation, not just the large ones. Also, the default formatting, e.g. `5e3` is problematic, not many people understand that it means `5 × 10^3` (we could rewrite that, but it's another potential can of worms).
- Yes the ideal fix is probably that it outputs `1,234 trillion` as it currently does.
- `thousand trillion` is worse than what we have now (which is no double prefixes until we reach `million trillion`, which in itself is a bug we wanted to fix when eventually there's an important enough use case)

This issue makes me wonder if we want to rely on a 3rd party library to this degree. It's possible there will be another feature in the future that is harder to implement because we have to circumvent `d3-format`, and maybe it would force us to re-implement a subset of `d3-format` ourselves? What do you think?

We could also maybe fork `d3-format` but then we'd have to maintain a codebase which we're not using large parts of.